### PR TITLE
feat: improve index hits

### PR DESCRIPTION
**Pull request by zpw-123**
EdgeCoreTest.java与VertexCoreTest.java是部分索引的例子！

## Purpose of the PR

- close #xxx  

## Verifying these changes

- [ ] Trivial rework / code cleanup without any test coverage. (No Need)
- [ ] Already covered by existing tests, such as *(please modify tests here)*.
- [ ] Need tests and can be verified as follows:
    - xxx

## Does this PR potentially affect the following parts?

- [ ]  Nope
- [ ]  Dependencies (add/update license info) 
- [ ]  Modify configurations
- [ ]  The public API
- [ ]  Other affects (typed here)

## Documentation Status

- [ ]  `Doc - TODO` 
- [ ]  `Doc - Done` 
- [ ]  `Doc - No Need`

### CODE REVIEW

File: hugegraph-core/src/main/java/org/apache/hugegraph/backend/tx/GraphIndexTransaction.java [hunk 1]
```diff
@@ -1554,20 +1560,31 @@ protected void removeIndex(IndexLabel indexLabel) {
...
+        private final Set<SchemaLabel> schemaLabels;
+        private final Set<IndexLabel> indexLabels;
 
-        private SchemaLabel schemaLabel;
-        private Set<IndexLabel> indexLabels;
 
         public MatchedIndex(SchemaLabel schemaLabel,
                             Set<IndexLabel> indexLabels) {
-            this.schemaLabel = schemaLabel;
+            //this.schemaLabels = new HashSet<>();
```
Comment (javeme) [inline hugegraph-core/src/main/java/org/apache/hugegraph/backend/tx/GraphIndexTransaction.java:1571]: remove it?

File: hugegraph-core/src/main/java/org/apache/hugegraph/backend/tx/GraphTransaction.java [hunk 2]
```diff
@@ -1395,8 +1397,8 @@ private <R> QueryList<R> optimizeQueries(Query query,
         }
 
         boolean supportIn = this.storeFeatures().supportsQueryWithInCondition();
-        for (ConditionQuery cq : ConditionQueryFlatten.flatten(
-                                 (ConditionQuery) query, supportIn)) {
+        for (ConditionQuery cq: ConditionQueryFlatten.flatten(
+            (ConditionQuery) query, supportIn)) {
```
Comment (javeme) [inline hugegraph-core/src/main/java/org/apache/hugegraph/backend/tx/GraphTransaction.java:1401]: prefer to keep the old style

File: hugegraph-core/src/main/java/org/apache/hugegraph/backend/tx/GraphTransaction.java [hunk 3]
```diff
@@ -1405,13 +1407,127 @@ private <R> QueryList<R> optimizeQueries(Query query,
              * 2.index-query result(ids after optimization), which may be empty.
              */
             if (q == null) {
-                queries.add(this.indexQuery(cq), this.batchSize);
+                boolean sys = cq.syspropConditions().size() != 0;
```
Comment (javeme) [inline hugegraph-core/src/main/java/org/apache/hugegraph/backend/tx/GraphTransaction.java:1410]: sysprop?

File: hugegraph-core/src/main/java/org/apache/hugegraph/backend/tx/AbstractTransaction.java [hunk 4]
```diff
@@ -133,7 +133,7 @@ public Number queryNumber(Query query) {
     }
 
     @Watched(prefix = "tx")
-    public QueryResults<BackendEntry> query(Query query) {
+    protected QueryResults<BackendEntry> query(Query query) {
```
Comment (javeme) [inline hugegraph-core/src/main/java/org/apache/hugegraph/backend/tx/AbstractTransaction.java:136]: not sure why need to mark protected

File: hugegraph-core/src/main/java/org/apache/hugegraph/backend/tx/GraphTransaction.java [hunk 5]
```diff
@@ -1405,13 +1407,127 @@ private <R> QueryList<R> optimizeQueries(Query query,
...
+            if(this.graph().propertyKey(pk).olap()) {
+                return true;
+            }
+        }
+
+        return false;
+    }
+    // Method to reorder conditions
+    private void excludeOnlyLabelQuery(ConditionQuery cq, QueryList queries) {
+        // 判断是否命中索引
```
Comment (javeme) [inline hugegraph-core/src/main/java/org/apache/hugegraph/backend/tx/GraphTransaction.java:1434]: can we translate all Chinese comments?

File: hugegraph-core/src/main/java/org/apache/hugegraph/backend/tx/GraphTransaction.java [hunk 6]
```diff
@@ -1405,13 +1407,127 @@ private <R> QueryList<R> optimizeQueries(Query query,
...
+        Collection<VertexLabel> vertexLabels = this.graph().vertexLabels();
+        Collection<EdgeLabel> edgeLabels = this.graph().edgeLabels();
+
+        Set<SchemaLabel> matchedLabels =
+            matchedIndices.stream()
+                .flatMap(matchedIndex -> matchedIndex.schemaLabels().stream())
+                .collect(Collectors.toSet());
+
+        Id label = rawQuery.condition(HugeKeys.LABEL);
+        if (label == null) {
```
Comment (javeme) [inline hugegraph-core/src/main/java/org/apache/hugegraph/backend/tx/GraphTransaction.java:1475]: prefer `return if (label != null)`

**Review by javeme — COMMENTED** [review]
Thank you very much for your PR.
Could you briefly describe the feature intent and design framework?

**Comment by simon824** [comment]
hi @zpw-123 , is it still in progress?

**Comment by zpw-123** [comment]
是的



---Original---
From: "Simon ***@***.***&gt;
Date: Fri, Oct 27, 2023 09:51 AM
To: ***@***.***&gt;;
Cc: ***@***.******@***.***&gt;;
Subject: Re: [apache/incubator-hugegraph] feat: improve index hits (PR #2314)




 
hi @zpw-123 , is it still in progress?

**Pull request closed without merging**